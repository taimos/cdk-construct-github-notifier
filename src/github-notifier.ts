import * as fs from 'fs';
import * as path from 'path';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as events from '@aws-cdk/aws-events';
import * as targets from '@aws-cdk/aws-events-targets';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as lambdajs from '@aws-cdk/aws-lambda-nodejs';
import * as cdk from '@aws-cdk/core';

export interface GithubNotifierProps {
  readonly githubToken: cdk.SecretValue;
}

export class GithubNotifier extends cdk.Construct {

  private readonly eventHandler: lambdajs.NodejsFunction;

  constructor(scope: cdk.Construct, id: string, props: GithubNotifierProps) {
    super(scope, id);

    const entry = fs.existsSync(path.join(__dirname, 'lambda', 'index.ts'))
      ? path.join(__dirname, 'lambda', 'index.ts') // local development
      : path.join(__dirname, 'lambda', 'index.js'); // when published in npm

    this.eventHandler = new lambdajs.NodejsFunction(this, 'Resource', {
      entry,
      runtime: lambda.Runtime.NODEJS_12_X,
      timeout: cdk.Duration.seconds(5),
      environment: {
        GITHUB_TOKEN: props.githubToken.toString(),
      },
      bundling: {
        minify: true,
      },
    });
    this.eventHandler.addToRolePolicy(new iam.PolicyStatement({
      actions: ['codepipeline:GetPipelineExecution'],
      resources: ['*'],
    }));
  }

  public onPipelineStateChange(pipeline: codepipeline.Pipeline, stateContext?: string) {
    pipeline.onStateChange('PipelineStateChange', {
      target: new targets.LambdaFunction(this.eventHandler, {
        event: events.RuleTargetInput.fromObject({
          event: events.EventField.fromPath('$'),
          context: {
            stateContext: stateContext ?? 'CodePipeline',
          },
        }),
      }),
    });
  }

  public onStageStateChange(stage: codepipeline.IStage, stateContext?: string) {
    stage.onStateChange('StageStateChange', new targets.LambdaFunction(this.eventHandler, {
      event: events.RuleTargetInput.fromObject({
        event: events.EventField.fromPath('$'),
        context: {
          stateContext: stateContext ?? `CodePipeline-${stage.stageName}`,
        },
      }),
    }));
  }

  public onActionStateChange(action: codepipeline.IAction, stateContext: string) {
    action.onStateChange('ActionStateChange', new targets.LambdaFunction(this.eventHandler, {
      event: events.RuleTargetInput.fromObject({
        event: events.EventField.fromPath('$'),
        context: {
          stateContext,
        },
      }),
    }));
  }

}