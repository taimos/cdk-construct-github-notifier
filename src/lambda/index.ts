import * as AWS from 'aws-sdk';
import * as axios from 'axios';

const codepipeline = new AWS.CodePipeline();

interface GitHubPayload {
  state: string;
  'target_url': string;
  description: string;
  context: string;
}

export async function handler(targetEvent: any) {
  console.log(targetEvent);

  const region = targetEvent.event.region;
  const pipelineName = targetEvent.event.detail.pipeline;
  const executionId = targetEvent.event.detail['execution-id'];
  const state = transformState(targetEvent.event.detail.state);
  const githubContext = targetEvent.context.stateContext;

  if (state === null) {
    return;
  }

  const result = await getPipelineExecution(pipelineName, executionId);
  const payload = createPayload(pipelineName, region, state, githubContext);
  await postStatusToGitHub(result.owner, result.repository, result.sha, payload);

  return null;
};

function transformState(state: string) {
  if (state === 'STARTED' || state === 'CANCELED') {
    return 'pending';
  }
  if (state === 'SUCCEEDED') {
    return 'success';
  }
  if (state === 'FAILED' || state === 'SUPERSEDED') {
    return 'failure';
  }
  return null;
}

function createPayload(pipelineName: string, region: string, state: string, context: string): GitHubPayload {
  let description;
  if (state === 'pending') {
    description = 'Build started';
  } else if (state === 'success') {
    description = 'Build succeeded';
  } else if (state === 'failure') {
    description = 'Build failed!';
  } else {
    throw new Error('Invalid state');
  }

  return {
    state,
    target_url: buildCodePipelineUrl(pipelineName, region),
    description,
    context,
  };
}

function buildCodePipelineUrl(pipelineName: string, region: string) {
  return `https://${region}.console.aws.amazon.com/codepipeline/home?region=${region}#/view/${pipelineName}`;
}

async function getPipelineExecution(pipelineName: string, executionId: string) {
  const params = {
    pipelineName: pipelineName,
    pipelineExecutionId: executionId,
  };

  const result = await codepipeline.getPipelineExecution(params).promise();
  const artifactRevision = result.pipelineExecution!.artifactRevisions![0];

  const revisionURL = artifactRevision.revisionUrl!;
  const sha = artifactRevision.revisionId!;

  const pattern = /github.com\/(.+)\/(.+)\/commit\//;
  const matches = pattern.exec(revisionURL);

  return {
    owner: matches![1],
    repository: matches![2],
    sha: sha,
  };
};

async function postStatusToGitHub(owner: string, repository: string, sha: string, payload: GitHubPayload) {
  const url = `/${owner}/${repository}/statuses/${sha}`;

  const config: axios.AxiosRequestConfig = {
    baseURL: 'https://api.github.com/repos',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${process.env.GITHUB_TOKEN}`,
    },
  };

  await axios.default.post(url, payload, config);
};
