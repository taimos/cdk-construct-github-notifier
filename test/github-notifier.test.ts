import * as codepipeline from '@aws-cdk/aws-codepipeline';
import * as actions from '@aws-cdk/aws-codepipeline-actions';
import * as cdk from '@aws-cdk/core';
import * as testee from '../src';

test('snapshot of Github Notifier', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'TestStack');

  const notifier = new testee.GithubNotifier(stack, 'GH', {
    githubToken: cdk.SecretValue.plainText('DUMMY'),
  });

  const pipeline = new codepipeline.Pipeline(stack, 'Pipeline');

  const sourceArtifact = new codepipeline.Artifact();

  pipeline.addStage({
    stageName: 'Source',
    actions: [
      new actions.GitHubSourceAction({
        actionName: 'Source',
        oauthToken: cdk.SecretValue.plainText('DUMMY'),
        output: sourceArtifact,
        owner: 'dummy',
        repo: 'repo',
      }),
    ],
  });
  notifier.onPipelineStateChange(pipeline, 'TestMessage');

  const stage = pipeline.addStage({
    stageName: 'Deploy',
  });
  notifier.onStageStateChange(stage);

  const action = new actions.CloudFormationCreateUpdateStackAction({
    actionName: 'Deploy',
    adminPermissions: true,
    stackName: 'testStack',
    templatePath: sourceArtifact.atPath('test.json'),
  });
  stage.addAction(action);
  notifier.onActionStateChange(action, 'DeployAction');

  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});