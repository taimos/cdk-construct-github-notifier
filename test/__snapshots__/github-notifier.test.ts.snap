// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot of Github Notifier 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersd84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5ArtifactHash35D2F059": Object {
      "Description": "Artifact hash for asset \\"d84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5\\"",
      "Type": "String",
    },
    "AssetParametersd84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5S3Bucket8CBA02B6": Object {
      "Description": "S3 bucket for asset \\"d84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5\\"",
      "Type": "String",
    },
    "AssetParametersd84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5S3VersionKey60BE9AC5": Object {
      "Description": "S3 key for asset version \\"d84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "GHCEC69A31": Object {
      "DependsOn": Array [
        "GHServiceRoleDefaultPolicy61907956",
        "GHServiceRoleFEA1E84C",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersd84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5S3Bucket8CBA02B6",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5S3VersionKey60BE9AC5",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersd84f0aa33cec61e718a0c2bf918ea4dc339bcae891f1f192514e34dd49a984c5S3VersionKey60BE9AC5",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "GITHUB_TOKEN": "DUMMY",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GHServiceRoleFEA1E84C",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 5,
      },
      "Type": "AWS::Lambda::Function",
    },
    "GHServiceRoleDefaultPolicy61907956": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "codepipeline:GetPipelineExecution",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GHServiceRoleDefaultPolicy61907956",
        "Roles": Array [
          Object {
            "Ref": "GHServiceRoleFEA1E84C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GHServiceRoleFEA1E84C": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineArtifactsBucket22248F97": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "KMSMasterKeyID": Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucketEncryptionKey01D58D69",
                    "Arn",
                  ],
                },
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "PipelineArtifactsBucketEncryptionKey01D58D69": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "PipelineRoleD68726F7",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "PipelineDeployRole97597E3E",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::GetAtt": Array [
                    "PipelineDeployCodePipelineActionRole8B83082E",
                    "Arn",
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "PipelineArtifactsBucketEncryptionKeyAlias5C510EEE": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AliasName": "alias/codepipeline-teststackpipeline28b9b4c0",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "PipelineArtifactsBucketEncryptionKey01D58D69",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
      "UpdateReplacePolicy": "Delete",
    },
    "PipelineC660917D": Object {
      "DependsOn": Array [
        "PipelineRoleDefaultPolicyC7A05455",
        "PipelineRoleD68726F7",
      ],
      "Properties": Object {
        "ArtifactStore": Object {
          "EncryptionKey": Object {
            "Id": Object {
              "Fn::GetAtt": Array [
                "PipelineArtifactsBucketEncryptionKey01D58D69",
                "Arn",
              ],
            },
            "Type": "KMS",
          },
          "Location": Object {
            "Ref": "PipelineArtifactsBucket22248F97",
          },
          "Type": "S3",
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "PipelineRoleD68726F7",
            "Arn",
          ],
        },
        "Stages": Array [
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Source",
                  "Owner": "ThirdParty",
                  "Provider": "GitHub",
                  "Version": "1",
                },
                "Configuration": Object {
                  "Branch": "master",
                  "OAuthToken": "DUMMY",
                  "Owner": "dummy",
                  "PollForSourceChanges": false,
                  "Repo": "repo",
                },
                "Name": "Source",
                "OutputArtifacts": Array [
                  Object {
                    "Name": "Artifact_Source_Source",
                  },
                ],
                "RunOrder": 1,
              },
            ],
            "Name": "Source",
          },
          Object {
            "Actions": Array [
              Object {
                "ActionTypeId": Object {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CloudFormation",
                  "Version": "1",
                },
                "Configuration": Object {
                  "ActionMode": "CREATE_UPDATE",
                  "Capabilities": "CAPABILITY_NAMED_IAM",
                  "RoleArn": Object {
                    "Fn::GetAtt": Array [
                      "PipelineDeployRole97597E3E",
                      "Arn",
                    ],
                  },
                  "StackName": "testStack",
                  "TemplatePath": "Artifact_Source_Source::test.json",
                },
                "InputArtifacts": Array [
                  Object {
                    "Name": "Artifact_Source_Source",
                  },
                ],
                "Name": "Deploy",
                "RoleArn": Object {
                  "Fn::GetAtt": Array [
                    "PipelineDeployCodePipelineActionRole8B83082E",
                    "Arn",
                  ],
                },
                "RunOrder": 1,
              },
            ],
            "Name": "Deploy",
          },
        ],
      },
      "Type": "AWS::CodePipeline::Pipeline",
    },
    "PipelineDeployCodePipelineActionRole8B83082E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineDeployCodePipelineActionRoleDefaultPolicyEE6D615B": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineDeployRole97597E3E",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineArtifactsBucketEncryptionKey01D58D69",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "cloudformation:CreateStack",
                "cloudformation:DescribeStack*",
                "cloudformation:GetStackPolicy",
                "cloudformation:GetTemplate*",
                "cloudformation:SetStackPolicy",
                "cloudformation:UpdateStack",
                "cloudformation:ValidateTemplate",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":cloudformation:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":stack/testStack/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineDeployCodePipelineActionRoleDefaultPolicyEE6D615B",
        "Roles": Array [
          Object {
            "Ref": "PipelineDeployCodePipelineActionRole8B83082E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineDeployRole97597E3E": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "cloudformation.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineDeployRoleDefaultPolicy90429F22": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineArtifactsBucketEncryptionKey01D58D69",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "*",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineDeployRoleDefaultPolicy90429F22",
        "Roles": Array [
          Object {
            "Ref": "PipelineDeployRole97597E3E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelinePipelineStateChangeAllowEventRuleTestStackPipelinePipelineStateChangeDF7C3903290E2D85": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "GHCEC69A31",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "PipelinePipelineStateChangeBDC73521",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "PipelinePipelineStateChangeBDC73521": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail-type": Array [
            "CodePipeline Pipeline Execution State Change",
          ],
          "resources": Array [
            Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":codepipeline:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":",
                  Object {
                    "Ref": "PipelineC660917D",
                  },
                ],
              ],
            },
          ],
          "source": Array [
            "aws.codepipeline",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "GHCEC69A31",
                "Arn",
              ],
            },
            "Id": "Target0",
            "InputTransformer": Object {
              "InputPathsMap": Object {
                "f1": "$",
              },
              "InputTemplate": "{\\"event\\":<f1>,\\"context\\":{\\"stateContext\\":\\"TestMessage\\"}}",
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "PipelineRoleD68726F7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codepipeline.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "PipelineRoleDefaultPolicyC7A05455": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "PipelineArtifactsBucket22248F97",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "PipelineArtifactsBucket22248F97",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:ReEncrypt*",
                "kms:GenerateDataKey*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineArtifactsBucketEncryptionKey01D58D69",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "PipelineDeployCodePipelineActionRole8B83082E",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PipelineRoleDefaultPolicyC7A05455",
        "Roles": Array [
          Object {
            "Ref": "PipelineRoleD68726F7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PipelineSourceWebhookResourceD1089B44": Object {
      "Properties": Object {
        "Authentication": "GITHUB_HMAC",
        "AuthenticationConfiguration": Object {
          "SecretToken": "DUMMY",
        },
        "Filters": Array [
          Object {
            "JsonPath": "$.ref",
            "MatchEquals": "refs/heads/{Branch}",
          },
        ],
        "RegisterWithThirdParty": true,
        "TargetAction": "Source",
        "TargetPipeline": Object {
          "Ref": "PipelineC660917D",
        },
        "TargetPipelineVersion": 1,
      },
      "Type": "AWS::CodePipeline::Webhook",
    },
  },
}
`;
